### PR TITLE
pv e2e refactor and pre-bind test

### PR DESCRIPTION
**Pull request by jeffvance**
refactored persistentvolume e2e so that multiple It() tests can be run. Added one test case for pre-binding, but the overall structure of the test should allow additional test cases to be more easily added.

### CODE REVIEW

**Comment by jeffvance** [comment]
@copejon @jsafrane PTAL. I'm leaving for vacation from Wed - Fri of this week. But, maybe @copejon can answer questions...

**Comment by jsafrane** [comment]
@k8s-bot ok to test

File: test/e2e/persistent_volumes.go [hunk 1]
```diff
@@ -22,149 +22,230 @@ import (
...
+		}
+	})
 
-		// Create the nfs server pod
-		pod := startVolumeServer(c, config)
-		serverIP := pod.Status.PodIP
-		framework.Logf("NFS server IP address: %v", serverIP)
+
+	// Individual tests follow:
+	It("should create a PersistentVolume, Claim, and a client Pod that will test the read/write access of the volume", func() {
```
Comment (copejon) [inline test/e2e/persistent_volumes.go]: Missing [Flaky].

Should read "...will test write access of the volume"

File: test/e2e/persistent_volumes.go [hunk 2]
```diff
@@ -22,155 +22,358 @@ import (
...
-		pvc, err = c.PersistentVolumeClaims(ns).Get(pvc.Name)
-		if pvcPhase := pvc.Status.Phase; pvcPhase != "Bound" {
-			framework.Failf("Expected PersistentVolumeClaim status Bound. Actual:  %+v.  Error: %+v", pvcPhase, err)
-		}
+	// Individual tests follow:
+	//
+	// Create an nfs PV, a claim that matches the PV, a pod that contains the
+	// claim. Verify that the PV and PVC bind correctly and that the pod can
+	// write to the nfs volume.
+	It("should create a PersistentVolume, Claim, and Pod that will test write access of the volume", func() {
```
Comment (copejon) [inline test/e2e/persistent_volumes.go]: @jeffvance Unless we can show that our fixes correct the flakiness this test cause, I think we still need the [Flaky] string at the end of the It statement.

**Comment by jsafrane** [comment]
`gofmt` check failed:
```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s -w' needs to be run on the following files: 
./test/e2e/persistent_volumes.go
FAILED   ./hack/../hack/verify-gofmt.sh	9s
```

**Comment by jsafrane** [comment]
Please rebase, @copejon's "Overhauled pv e2e test to reflect common lifecycle and de-flake" patch is already merged

File: test/e2e/persistent_volumes.go [hunk 3]
```diff
@@ -47,16 +47,158 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
...
+	}
+}
+
+// create the PV resource. Fails test on error.
+func createPV(c *client.Client, pv *api.PersistentVolume) (*api.PersistentVolume) {
+
+	pv, err := c.PersistentVolumes().Create(pv)
+	if err != nil {
+		framework.Failf("Create PersistentVolume %v failed: %v", pv.Name, err)
+		return nil
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: `return nil` suggests, that the execution after `Failf` continues, which is IMO not the case. Please remove it (and also few lines below in `createPVC`.

File: test/e2e/persistent_volumes.go [hunk 4]
```diff
@@ -47,16 +47,158 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
...
+
+	if pvFirst {
+		By("Creating the PV followed by the PVC")
+		pv = createPV(c, pv)
+		pvc = createPVC(c, ns, pvc)
+	} else {
+		By("Creating the PVC followed by the PV")
+		pvc = createPVC(c, ns, pvc)
+		// need to fill-in claimRef with pvc.Name 
+		pv.Spec.ClaimRef.Name = pvc.Name
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: why do you "need to fill-in claimRef with pvc.Name"? It's either filled by `makePersistentVolume` call above or the PV should not be pre-bound.

File: test/e2e/persistent_volumes.go [hunk 5]
```diff
@@ -47,16 +47,158 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
...
+
+	// Wait for pv and pvc to bind to each other
+	WaitOnPVandPVC(c, ns, pv, pvc)
+
+	// Check that the PersistentVolume.ClaimRef is valid and matches the PVC
+	framework.Logf("Checking PersistentVolume ClaimRef is non-nil")
+	pv, err := c.PersistentVolumes().Get(pv.Name)
+	if err != nil {
+		framework.Failf("Cannot re-get PersistentVolume %v:", pv.Name, err)
+		return nil, nil
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: again, remove `return` after `Failf`

File: test/e2e/persistent_volumes.go [hunk 6]
```diff
@@ -208,50 +318,62 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
...
-		pv, err = c.PersistentVolumes().Get(pv.Name)
-		if pv.Spec.ClaimRef != nil && len(pv.Spec.ClaimRef.UID) > 0 {
-			crjson, _ := json.MarshalIndent(pv.Spec.ClaimRef, "", "  ")
-			framework.Failf("Expected a nil pv.ClaimRef or empty UID. Found: ", string(crjson))
-		}
-
-		// Delete the PersistentVolume
-		By("Deleting PersistentVolume")
+		// Last cleanup step is to delete the pv
 		deletePersistentVolume(c, pv)
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: As the tests use `Failf` a lot, any created PV will be leaked after failed test. You should call `defer deletePersistentVolume(c, pv)` just after `createPVandPVC` (+ the same in the test below).

**Comment by copejon** [comment]
@jeffvance re: gofmt - Dev docs mention setting up a git hook that will fire off when you commit.  It'll run a series of checks (include gofmt) for you 

see: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#committing-changes-to-your-fork

**Comment by jsafrane** [comment]
@k8s-bot ok to test

**Comment by jeffvance** [comment]
@jsafrane @copejon Thanks for the review! I'm at Red Hat Summit. I made the suggested changes before the conference but cannot test due to a just-broken laptop... Will resume Fri.

File: test/e2e/persistent_volumes.go (hunk 4 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: @jsafrane pv.ClaimRef.name is not filled in (by makePersistentVolume) when the pv is created because the associated claim does not know its own name until the claim is instantiated. I believe this is due to using a generated named.

File: test/e2e/persistent_volumes.go (hunk 6 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: I believe that Failf() triggers AfterEach's so I coded the AfterEach to delete the pod, pvc, and pv if their values are not nil.  Does this approach solve your comment?

File: test/e2e/persistent_volumes.go (hunk 4 above)
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: I see. This is not obvious from the code, could you please add a comment that would explain that pvc.Name is generated?

File: test/e2e/persistent_volumes.go (hunk 6 above)
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: Ok, I missed that.

File: test/e2e/persistent_volumes.go (hunk 4 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: I added additional comments regarding generated names.

**Comment by jeffvance** [comment]
@jsafrane @copejon I've pushed up changes based on your comments. PTAL. @pmorie If you see LGTMs can you please merge? I have more PV test cases I want to add and this PR is the foundation for that. Thanks.

**Comment by jsafrane** [comment]
```
!!! 'gofmt -s -w' needs to be run on the following files: 
./test/e2e/persistent_volumes.go
FAILED   ./hack/../hack/verify-gofmt.sh	9s
```

**Comment by jeffvance** [comment]
More refactoring and minor change to nfs.go. The test is still [Flaky]. I get a pod nfs mount failure about 35% of the time on the test case, It(), whichever is run last. This failure prevents the pod from running and, when the pv is deleted in the AfterEach, the mount problem also causes the recycler pod's mount to fail. When the test exits the recycler's nfs mount on the host is present and must be `umount`'d. And the files in _/var/lib/kubelet/pods/*_ should be removed. I do not yet know if the mount problem is a result of my changes to the test, or due to my laptop environment, or something else. Any thoughts here are much appreciated...

**Comment by jsafrane** [comment]
Please, pay attention to test results:

```
test/e2e/persistent_volumes.go:115: wrong number of args for format in Errorf call: 1 needed but 2 args
test/e2e/persistent_volumes.go:234: wrong number of args for format in Errorf call: 1 needed but 2 args
test/e2e/persistent_volumes.go:239: wrong number of args for format in Errorf call: 1 needed but 2 args
exit status 1
!!! Error in ./hack/../hack/verify-govet.sh:49
```

And regarding leftover mount... I saw it couple of times, I have no explanation though. Just brainstorming: can  it happen that the NFS server pod is destroyed while the volume it provides is still mounted to a pod that is (slowly) being destroyed at the same time? Kubelet will then time out unmounting the client. Does it help if you sleep few seconds after destroying the client pod and before destroying the NFS server pod?

**Comment by jeffvance** [comment]
@jsafrane That compile error above is very strange. I didn't push code that didn't compile.
Re. potential timing windows: sometimes the test (both It() cases) runs with no errors and no stale mounts leftover on the host. Other times the both test succeed but the last test case leaves the nfs recycler nfs mount still mounted on the host, and thus the _/var/lib/kubelet/pods/*_ files remain. And, some runs of the test succeed but leave 3-4 stale nfs mounts on the host and 5-6 stale secret mounts on the host. All of these issues could be timing related, maybe. I do clean up the host consistently between consecutive runs of the test.
I would like to discover why this test is [Flaky] and fix those problems so that it can be run in the mainstream tests, but so far I have not been able to.

File: pkg/volume/nfs/nfs.go [hunk 7]
```diff
@@ -237,24 +237,24 @@ func (c *nfsUnmounter) TearDown() error {
...
 	}
 	notMnt, mntErr := c.mounter.IsLikelyNotMountPoint(dir)
 	if mntErr != nil {
-		glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+		glog.Errorf("nfs: IsLikelyNotMountPoint check failed: %v", mntErr)
 		return mntErr
 	}
 	if notMnt {
-		if err := os.Remove(dir); err != nil {
+		if err := os.RemoveAll(dir); err != nil {
```
Comment (saad-ali) [inline pkg/volume/nfs/nfs.go]: Curious, what motivated this change. After a volume is unmounted the directory should already be empty right?

Comment (saad-ali) [inline pkg/volume/nfs/nfs.go]: My fear is that this may mask a real bug where NFS volume did not unmount

**Comment by saad-ali** [comment]
I'm going to have to defer the PV/PVC test changes CR to @jsafrane since he is much more familiar with the PV/PVC logic.

File: pkg/volume/nfs/nfs.go (hunk 7 above)
Comment (jeffvance) [inline pkg/volume/nfs/nfs.go]: @saad-ali Thank you for your review. I made the couple of changes to nfs.go to mirror the glusterfs plugin where, 1) the plugin name, "nfs" appears in log messages, and 2) os.RemoveAll is called instead of just os.Remove. I agree with your comment that the dir should be empty and that it is more destructive to RemoveAll. Would not the same argument apply to all shareable volumes, eg. apply to gluster too? I have no problem leaving in os.Remove.

Comment (saad-ali) [inline pkg/volume/nfs/nfs.go]: Yes ideally they should be consistent. I'll leave it up to you to decide, but if there is not a bug that is being fixed, then I'd favor leaving it as is.

**Comment by jeffvance** [comment]
@saad-ali I changed the RemoveAll() calls in nfs.go back to Remove() since there is no issue I am aware of in the nfs plugin where Remove() is not working correctly.

**Comment by childsb** [comment]
@jeffvance it seems like this is still failing build or test.  also the removeAll(...) calls are still in nfs.go.. did you forget to git push changes?

**Comment by jeffvance** [comment]
Sorry @childsb. I had pushed the changes but forgot `-f` and, unfortunately, didn't notice the push failure. It's there now.

**Comment by childsb** [comment]
why are you force pushing @jeffvance ?!!  you should avoid that unless doing a rebase

**Comment by jeffvance** [comment]
I do `git push origin <branch> -f` after doing a `git commit --amend`. I do the --amend to not create a new commit record and just reuse the previous text. But when I do --amend I need to force the push. Is there a better way?

**Comment by childsb** [comment]
thats a good reason to force push.  usually the workflow is to leave the commits in place so there's a record of before/after changes then squash commits at the end https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md#7-fix-feedback-in-a-new-commit

**Comment by jeffvance** [comment]
@saad-ali I think the size label is wrong on this pr. Though there is some significant refactoring, it is only one file. Maybe the size label doesn't matter, but my opinion is this pr is a size=med or size=small, FWIW.

**Comment by jeffvance** [comment]
@childsb Is there something I can do to get the 2 pending tests to run?

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #IGNORE

File: pkg/volume/nfs/nfs.go [hunk 8]
```diff
@@ -237,20 +237,20 @@ func (c *nfsUnmounter) TearDown() error {
 func (c *nfsUnmounter) TearDownAt(dir string) error {
 	notMnt, err := c.mounter.IsLikelyNotMountPoint(dir)
 	if err != nil {
-		glog.Errorf("Error checking IsLikelyNotMountPoint: %v", err)
+		glog.Errorf("nfs: Error checking IsLikelyNotMountPoint: %v", err)
```
Comment (jsafrane) [inline pkg/volume/nfs/nfs.go]: this belongs to a separate PR.

File: test/e2e/persistent_volumes.go [hunk 9]
```diff
@@ -18,159 +18,481 @@ package e2e
...
+	framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 90*time.Second)
+
+	// Double check that the PV is really deleted. It is important to not
+	// delete the nfs-server pod until after the pv is deleted/recycled.
+	pv, err = c.PersistentVolumes().Get(pv.Name)
+	if err == nil {
+		return pv, fmt.Errorf("PV %v deleted yet Get() succeeded", pv.Name)
+	}
+	if !apierrs.IsNotFound(err) {
+		return pv, fmt.Errorf("Get on deleted PV %v failed with error other than \"not found\" %v", pv.Name, err)
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: printing `err` is useless, we know it's nil

File: test/e2e/persistent_volumes.go [hunk 10]
```diff
@@ -18,159 +18,481 @@ package e2e
...
+	// to make sure the PVC gets deleted.
+	time.Sleep(5 * time.Second)
+
+	// Double check that the PVC is really deleted.
+	pvc, err = c.PersistentVolumeClaims(ns).Get(pvc.Name)
+	if err != nil && !apierrs.IsNotFound(err) {
+		return pv, pvc, fmt.Errorf("Get on deleted PVC %v failed with error other than \"not found\": %v", pvc.Name, err)
+	}
+	if err == nil {
+		return pv, pvc, fmt.Errorf("PVC %v deleted yet still exists: %v", pvc.Name, err)
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: printing `err` is useless, we know it's nil

File: test/e2e/persistent_volumes.go [hunk 11]
```diff
@@ -18,159 +18,481 @@ package e2e
...
+
+	By("Deleting PersistentVolumeClaim to trigger PV Recycling")
+
+	framework.Logf("Deleting PersistentVolumeClaim %v to trigger PV Recycling", pvc.Name)
+	err := c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+	if err != nil {
+		return pv, pvc, fmt.Errorf("Delete of PVC %v failed: %v", pvc.Name, err)
+	}
+
+	// There is no f.WaitForPersistentVolumeClaimDelete() so just sleep some
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: When `c.PersistentVolumeClaims(ns).Delete(pvc.Name)` returns, the PVC should be gone even without sleeping.

And if you need `f.WaitForPersistentVolumeClaimDelete` then make one, sleeping is suboptimal.

File: test/e2e/persistent_volumes.go [hunk 12]
```diff
@@ -18,159 +18,481 @@ package e2e
...
-		framework.Failf("Delete PersistentVolume failed: %v", err)
+		return pv, fmt.Errorf("Delete() PersistentVolume %v failed: %v", pv.Name, err)
+	}
+
+	// Wait for PersistentVolume to delete
+	framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 90*time.Second)
+
+	// Double check that the PV is really deleted. It is important to not
+	// delete the nfs-server pod until after the pv is deleted/recycled.
+	pv, err = c.PersistentVolumes().Get(pv.Name)
```
Comment (jsafrane) [inline test/e2e/persistent_volumes.go]: I don't think this check is necessary, it does the same thing as `WaitForPersistentVolumeDeleted` called above. Or did you experence `WaitForPersistentVolumeDeleted` returning no error when a PV still existed?

Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: @jsafrane Thank you for all of your comments today. Yes, I found that the wait timout was insufficient sometimes in my environment (laptop, all-in-one k8s cluster). This extra test to ensure the pod is gone does not hurt any does it?

File: test/e2e/persistent_volumes.go (hunk 9 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: If your comment is for line 73 above, we know that err != nil && err != <IsNotFound>.

File: test/e2e/persistent_volumes.go (hunk 10 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Agreed

File: test/e2e/persistent_volumes.go (hunk 11 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: removed the sleep.

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #IGNORE

Why does it always end with "Jenkins GCE Node e2e — Waiting for status to be reported"?

**Comment by jsafrane** [comment]
LGTM

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #IGNORE
